Connected pre-loaded images to v2

stephaniequintana opened the pull request:
<!-- Add a short description about your changes here-->
Allows the user to select a pre-loaded image to explore the tool with.

Fixes the "Connect Sample Images" from Planning issue #0415


![](https://media.giphy.com/media/qvih44rab9HyMV1qh2/giphy.gif)

### Caveats:
- the index2.html includes the code from #0435 which corrects keyboard-accessibility of the `<input>` element (see question in comments)
- the function for uploading the selected sample image has ONLY been added to the dist/inframgram2.js file 
  - it has not beeen added to the src/ui/interface.js file 
  - Q: how do we proceed with this? Create an interface2.js file for this version? Please advise.



* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updation
* [x] ask `@publiclab/reviewers` for help, in a comment below

<!--We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!-->

<!--If tests do fail, click on the red `X` to learn why by reading the logs.-->

<!--Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software -->

<!--Thanks!-->

stephaniequintana commented:
@jywarren, @TildaDares, @cesswairimu 

Clarifications/Discussion requested:
- In addition to not having added the function of this PR to the interface.js file (or having created an interface2.js file), I realize that the functions altered in the-previously-merged-PR #0431 (namely changing multiple occurances of `$('#preset-modal').modal('show');` to `$('#preset-modal').offcanvas('show');`) were also only added to the infragram2.js file and NOT to the interface.js file. How might we best handle these functions that only pertain to this version?
- It does not seem correct to have added the code from another PR into this PR...Is it correct that NOT including it in this PR would just create a simple merge-conflict? I believe I've confused myself into thinking that by not including that code into this PR then if that PR were merged first, it would would be overwritten when this PR is merged. I realize this question is poorly articulated, no need to reply if it's not clear :/

I hope this finds you all well :)

jywarren commented:
Hi, no issue, it's confusing! Merging will incorporate only whats shown in the changed files here, and the merge action always prioritizes the active PR's changes. You should just before merging be able to see in the diff view the green + lines that will be the final state after the merge is conducted. Once a merge is completed, the diff view and changed files in all other PRs will be modified to show the new post-merge changes taking into account the action that just happened. Does that make sense?

It's OK to include changes in both - it's very likely they'll then be dropped from the diff of the other, as that change will already have occurred, OR it may say there's a conflict and it can't be merged automatically.

jywarren commented:
> functions altered in the-previously-merged-PR https://github.com/publiclab/infragram/pull/431 (namely changing multiple occurances of $('#preset-modal').modal('show'); to $('#preset-modal').offcanvas('show');) were also only added to the infragram2.js file and NOT to the interface.js file. How might we best handle these functions that only pertain to this version?

Should we try this idea of detecting what version we're running? I'll make a code suggestion so the flag will exist!

jywarren commented:
I started the flag implementation in https://github.com/publiclab/infragram/pull/438 - i will try to generate the dist files too -- but if that makes sense we can merge it and you can begin using `if (options.version == 1) {}` in your code!

stephaniequintana commented:
@jywarren, 
I included the function in the `$(document).ready(function(){...}` of the src/ui/interface.js. 

Please ensure that the function was placed in the correct location.

I've also include it in the dist/infragram.js file in preparation of dist/infragram2.js becoming obsolete. The code does not break the original version and thus I did not utilize the version flag (implemented in #0438).

You can view this in action at [https://stephaniequintana.github.io/infragram/index2.html](https://stephaniequintana.github.io/infragram/index2.html)

jywarren commented on src/ui/interface.js:
```diff
@@ -68,6 +68,21 @@ module.exports = function Interface(options) {
       return true;
     });
 
+    $("#sample-image__select").click(function(e){
```
This looks good; it'll only run if there is an HTML DOM element with an id of `sample-image__select`, so this won't break v1. However, 2 thoughts - 

1. should we mark this as a v2 feature with a comment, so people know, and if we ever clean up all v1 code we can tell it apart
2. should we otherwise explain this briefly in a comment, since it sounds a little mysterious?

jywarren replied:
What's cool is that this means ANY image could be used as a sample image, it just has to have the id `sample-image__select`. I suggest that we change that to a class, since strictly, ids should be used only once. Then someday someone may want to add, for example, a bunch of different examples in a kind of menu, they could use this. So the class could be maybe... `.sample-image`?

jywarren replied:
But perhaps that means we should then apply the class to just the individual images, rather than the whole container. Let me make a suggestion!

jywarren commented on src/ui/interface.js:
```diff
@@ -68,6 +68,21 @@ module.exports = function Interface(options) {
 
+    $("#sample-image__select").click(function(e){
+      e.stopPropagation();
+      const img = (e.target.classList.contains('rfi')) ? document.getElementById('rfi') : document.getElementById('bfi');
```
Here's another super great line of code - i love that the idea is that you can add the "bfi" or "rfi" classes to switch between. Maybe we should explain this too in a comment!

jywarren replied:
Aha - i see now -- we want to click on the thumbnail, but use the big hidden image! I wonder - if we use relatively low res sample images, like, 1024x768 or less, could we simplify things by just skipping thumbnails? JPGs don't weigh that much, you know? Then the logic here can be much simpler - we just display the big images with `style="width:200px;"` or something like that!

jywarren replied:
So on this line, if we stop using separate thumbnails, we can maybe just skip using red-filter or blue-filter classes completely. Since we'd just select whatever image, and think about what kind of filter later?

jywarren commented on index2.html:
```diff
@@ -277,13 +275,15 @@ <h2 style="font-size:1.25rem;">Analyze plant health. <br> Visualize photosynthes
-                  <p class="card-text" style="font-size:.625rem;padding-top:.5rem;"><small>Red filtered trees</small></p>
+              <div id="sample-image__select" class="d-flex justify-content-around">
+                <div id="red-filtered-img" class="rfi d-flex flex-column justify-content-center align-items-baseline">
+                  <img class="rfi" src="assets/red-filtered-trees-thumbnail.jpeg" alt="Red filtered image - select to explore the tool." style="width:5rem;height:auto;border:.5rem solild var(--image-container-bg);">
```
```suggestion
                  <img class="sample-image red-filter" src="assets/red-filtered-trees-thumbnail.jpeg" alt="Red filtered image - select to explore the tool." style="width:5rem;height:auto;border:.5rem solild var(--image-container-bg);">
```

I explain this idea below - and as to naming, I really like longer descriptive names because it often means we don't have to use as many comments!

jywarren commented on index2.html:
```diff
@@ -277,13 +275,15 @@ <h2 style="font-size:1.25rem;">Analyze plant health. <br> Visualize photosynthes
                 </div>
-                <div class="d-flex flex-column justify-content-center align-items-center">
+                <div id="blue-filtered-img" class="d-flex flex-column justify-content-center align-items-center">
                   <img src="assets/blue-filtered-plant-thumbnail.jpeg" alt="Blue filtered image - select to explore the tool." style="width:5rem;height:auto;">
```
```suggestion
                  <img class="sample-image blue-filter" src="assets/blue-filtered-plant-thumbnail.jpeg" alt="Blue filtered image - select to explore the tool." style="width:5rem;height:auto;">
```

jywarren commented on src/ui/interface.js:
```diff
@@ -68,6 +68,21 @@ module.exports = function Interface(options) {
+          const file = new File([blob], img.src, blob)
+          const fileInput = document.querySelector('input[type="file"]');
+          const dataTransfer = new DataTransfer();
+          dataTransfer.items.add(file);
```
Wow i really like this way of doing it. It literally selects the image, same as someone using it! That will mean that we don't have to support (or think about) 2 different means of adding images. Very cool.

jywarren reviewed (changes requested):
Hi @stephaniequintana it took me a moment to catch up to you understanding the code here but it looks great. I made some suggestions for simplicity, legibility and such. If you take up the suggestion of eliminating thumbnail images, we can further simplify the HTML too, removing the red-filter classes I had added and a few of the ones you added as well. 

Let me know what you think!

stephaniequintana commented:
@jywarren,

Great suggestions! Thank you 🙏 

I originally put the eventListener on the entire div in case a user clicks the text (instead of the image), but this came with having to add many classes to many elements. This way is MUCH cleaner and the function itself is more clear. for your quick reference, it is now:
```javascript
$(".sample-image").click(function(e){
          e.stopPropagation();
          const img = (e.target);
          fetch(img.src)
            .then(res => res.blob())
            .then(blob => {
              const file = new File([blob], img.src, blob)
              const fileInput = document.querySelector('input[type="file"]');
              const dataTransfer = new DataTransfer();
              dataTransfer.items.add(file);
              fileInput.files = dataTransfer.files;
              $(options.fileSelector).trigger("change");
            })
        });
```

I also deleted the thumbnails and replaced the images with smaller versions (width: 1024).

I am working on a better design for the welcome-container as a whole and will include this in a separate UI PR.

jywarren commented:
Sounds super!!!

On Thu, Aug 4, 2022, 5:46 PM Stephanie Quintana ***@***.***>
wrote:

> @jywarren <https://github.com/jywarren>,
>
> Great suggestions! Thank you 🙏
>
> I originally put the eventListener on the entire div in case a user clicks
> the text (instead of the image), but this came with having to add many
> classes to many elements. This way is MUCH cleaner and the function itself
> is more clear. for your quick reference, it is now:
>
> $(".sample-image").click(function(e){
>
>           e.stopPropagation();
>
>           const img = (e.target);
>
>           fetch(img.src)
>
>             .then(res => res.blob())
>
>             .then(blob => {
>
>               const file = new File([blob], img.src, blob)
>
>               const fileInput = document.querySelector('input[type="file"]');
>
>               const dataTransfer = new DataTransfer();
>
>               dataTransfer.items.add(file);
>
>               fileInput.files = dataTransfer.files;
>
>               $(options.fileSelector).trigger("change");
>
>             })
>
>         });
>
> I also deleted the thumbnails and replaced the images with smaller
> versions (width: 1024).
>
> I am working on a better design for the welcome-container as a whole and
> will include this in a separate UI PR.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/publiclab/infragram/pull/436#issuecomment-1205801491>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAF6J2YWVOA6CLSXRUKZLTVXQ24JANCNFSM54IQ4XXQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
